### PR TITLE
[feat] Server endpoint skeleton for global COVID-19 stats

### PR DESCRIPTION
**Pull request by advayDev1**
Defines a new endpoint for global COVID-19 stats.  Our use of a proxied backend, a static file we push once a day, etc. can be controlled behind this facade.  We will use a single API to get both global stats and local stats given location.

Test:
````
curl -i     -H 'Content-Type: application/json' -H 'Who-Client-ID: 00000000-0000-0000-0000-000000000000' -H 'Who-Platform: IOS' -X POST -d ''     http://localhost:8080/WhoService/getCaseStats

HTTP/1.1 200 OK
Content-Type: application/json
Server: Development/1.0
Date: Tue, 31 Mar 2020 20:22:24 GMT
Cache-Control: no-cache
Expires: Mon, 01 Jan 1990 00:00:00 GMT
Content-Length: 123

{"globalStats":{"lastUpdated":1585641600000,"cases":719758,"deaths":33673,"recoveries":-1,"attribution":"WHO"},"ttl":21600}
````

### CODE REVIEW

File: server/appengine/src/main/java/who/PublicWhoServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
+package who;
+
+// All responses are Client-independent.
+// No Client instance is available in this service.
+public class PublicWhoServiceImpl implements PublicWhoService {
+
+  @Override public GetGlobalStatsResponse getGlobalStats(Void request) {
+    // TODO: Configure cache response headers. RFC 2616 Section 9.5 says we can cache
+    // POST requests.  Or we can enable GET method in present.rpc.RpcFilter.
```
Comment (epicfaace) [inline server/appengine/src/main/java/who/PublicWhoServiceImpl.java]: I think it makes more sense (from an API perspective) to make this a GET method.

Comment (advayDev1) [reply server/appengine/src/main/java/who/PublicWhoServiceImpl.java]: It does, but RpcFilter forced POST right now.

File: api/who/who.proto [hunk 2]
```diff
@@ -4,12 +4,22 @@ package who;
...
 
   // The client should not pass a precise location.
   rpc putLocation(PutLocationRequest) returns (Void);
+
+  // TODO: Add location-specific stats.
+}
+
+// These responses are publicly cacheable and clients must
+// not send a client ID header.
+service PublicWhoService {
```
Comment (crazybob) [inline api/who/who.proto]: I'd prefer we keep things simple and just use one service for now. All rpcs should always be idempotent. Caching policies for individual rpcs will likely vary.

File: api/who/who.proto [hunk 3]
```diff
@@ -21,6 +31,21 @@ message PutLocationRequest {
...
 
+message GetGlobalStatsResponse {
+  CaseStats stats = 1;
+  string attribution = 2;
+}
+
+// Statistics in a single or aggregate jurisdiction.
+message CaseStats {
+  // msec since epoch.
+  int64 lastUpdated = 1;
```
Comment (crazybob) [inline api/who/who.proto]: I usually use uint64 for timestamps.

File: server/appengine/src/main/java/who/PublicRpcInterceptor.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
+package who;
+
+import present.rpc.ClientException;
+import present.rpc.RpcInterceptor;
+import present.rpc.RpcInvocation;
+
+/**
+ * For publicly-cacheable RPCs.
+ */
+public class PublicRpcInterceptor implements RpcInterceptor {
```
Comment (crazybob) [inline server/appengine/src/main/java/who/PublicRpcInterceptor.java]: We shouldn't need this class. Best to isolate logic to ClientInterceptor.

File: server/appengine/src/main/java/who/PublicWhoRpcFilter.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+import present.rpc.LoggingInterceptor;
+import present.rpc.RpcFilter;
+import present.rpc.RpcInterceptorChain;
+
+/**
+ * Configures RPC services for publicly-cacheable APIs.
+ */
+@Singleton public class PublicWhoRpcFilter extends RpcFilter  {
```
Comment (crazybob) [inline server/appengine/src/main/java/who/PublicWhoRpcFilter.java]: We don't need this class. For future reference, we need just one RpcFilter class for the whole app. You can call service() multiple times.

File: server/appengine/src/main/java/who/PublicWhoServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+    // TODO: Get from data source.
+    return new GetGlobalStatsResponse.Builder()
```
Comment (crazybob) [inline server/appengine/src/main/java/who/PublicWhoServiceImpl.java]: Present RPC is designed to be protocol independent. Perhaps we could add a ttl field to the response message?

File: api/who/who.proto [hunk 7]
```diff
@@ -21,6 +31,21 @@ message PutLocationRequest {
   double longitude = 3;
 }
 
+message GetGlobalStatsResponse {
```
Comment (crazybob) [inline api/who/who.proto]: Are these intended to be global stats? Guess we can add fields for local stats later?

Comment (advayDev1) [reply api/who/who.proto]: Yes global only.

File: server/appengine/src/main/java/who/PublicWhoServiceImpl.java (hunk 6 above)
Comment (advayDev1) [reply server/appengine/src/main/java/who/PublicWhoServiceImpl.java]: done

File: api/who/who.proto (hunk 2 above)
Comment (advayDev1) [reply api/who/who.proto]: Per discussion down to one api for both local and global stats, thus removing the utility of a publicly-cacheable API/service.

**Comment by advayDev1** [comment]
PTAL @crazybob - per our discussion, Benefits of public-caching are minor and outweighed by having a single roundtrip for both global and local (eventual) stats. Thus the single api no longer need be prevented from reading the client-id, removing the usefulness of two separate services.

File: api/who/who.proto [hunk 8]
```diff
@@ -21,6 +25,30 @@ message PutLocationRequest {
   double longitude = 3;
 }
 
+message GetCaseStatsResponse {
+  CaseStats globalStats = 1;
+
+  // TODO: Add local stats fields too, potentially at multiple administrative levels.
+
+  // Clients should not re-request stats until at least ttl seconds.
+  int64 ttl = 2;
```
Comment (crazybob) [inline api/who/who.proto]: I usually default to uint32 or 64 unless a number can be negative.

**Pull request merged**